Add old-channels command

williambl opened the pull request:
Locked to console channel.

williambl committed:
Add old channels command
Not yet finished, but working

williambl committed:
Add channel/category blacklisting to old-channels command

williambl committed:
Make old-channels ignore webhook messages
Untested

williambl committed:
Make timer task to check time since channels had their last message
Makes old-channels very fast.

williambl committed:
Add scheduler for ChannelMessageChecker

williambl committed:
Remove a debug statement

KiriCattus commented:
When tested we found out that this command tries to loop over the channels but just keeps on going and adding ones it has already gotten.

williambl commented:
It seems like it works now. Perhaps updating JDA fixed it.

williambl commented:
No, that can't be it - JDA isn't updated in this branch.

Whatever the reason, it seems to work.

KiriCattus commented:
I'll give it a test again on the larger server in a little bit, last I remember it just kept on looping through the server and getting duplication's of multiple channels

KiriCattus commented:
I've tested out the channels command again and threw in a few breakpoints on lines 46, 62 and 78 with and without lines 41 to 44 disabled, nothing happened in discord after the command ``!mmd-old-channels`` was run and nothing showed up in the logs or caught on the breakpoints.

williambl committed:
Finish merging

williambl committed:
Finalise old channels command
 - fix it throwing an exception if done too early
 - use the `Utils.checkCommand` method to check command
 - fix indentation
 - fix arg parsing

williambl commented:
Okay, this should be working fine now, and merge conflicts are fixed. Should be ready for merge after review.

![image](https://user-images.githubusercontent.com/5115825/104027228-fc180200-51be-11eb-918a-c72e8ad73855.png)

williambl commented on src/main/java/com/mcmoddev/mmdbot/updatenotifiers/forge/ForgeVersionHelper.java:
```diff
@@ -43,7 +44,7 @@ public static MinecraftForgeVersion getLatestMcVersionForgeVersions() throws Exc
     }
 
-    private static InputStreamReader openUrl() throws Exception {
+    private static InputStreamReader openUrl() throws IOException {
```
Didn't realise I'd committed this :|

williambl committed:
Rewrite ChannelMessageChecker to be asynchronous
All is streams. All is lambdas. I have looked into the `Future` and it is `Completable`.
All hail the asynchrony.

Co-authored-by: sciwhiz12 <arnoldnunag12@gmail.com>

williambl committed:
Replace oldchannels blacklist with whitelist

KiriCattus commented:
Suggestion to try with this, I noted with the relay bot and the mod we where using that the bot would continue to loop over all the channels in the entire server until I blocked the relay bot from being able to see the community channels (reducing the number of channels it could see therefore somehow stopping the bot looping and breaking) maybe we should limit this command to the community mods category since that is the main if not the only place it should be checking?

williambl commented:
I thought last we checked this worked?

KiriCattus commented:
I thought the last time I ran it locally and tested it that we still had channel looping but I can go and double check it now to be sure because I could be wrong, if I am I'll take that back and apologise for sure

williambl committed:
Add missing import

KiriCattus commented:
Update: we found the issues, listed them in discord last night, channel perm checks need to be in place or the timer fails entirely, if a perm is missing it should skip that channel as if it does not exist, category checking is broken (Or so Will says) channels can't be mentioned, only typed out for example ``proxys-trinkets`` not ``#proxys-trinkets`` but other than that things are all good :) my assumption was incorrect and I was remembering a different now resolved issue so my apologies :s

williambl committed:
Remove unneeded filter in old-channels command causing categories to not work

williambl committed:
Make old-channels command case-insensitive

KiriCattus commented:
If we can figure out that metals channel issue where it does not say when the last message was (currently it says 0 when the last message was a while ago) and add in the following we should be good for this PR :D
Missing channel perms catch/exception for ${channels}
Non-existing channel/category reply ${channel/category}
(Paste from Discord chat so we don't loose track like I did last time...)

jriwanek committed:
Merge branch '3.0' into old-channels
